src/test/java/test/configuration/issue165/AfterGroupsTest.java

```diff
@@ -0,0 +1,36 @@
+package test.configuration.issue165;
+
+import static org.testng.Assert.assertTrue;
+
+import java.util.function.Supplier;
+import org.testng.TestNG;
+import org.testng.annotations.DataProvider;
+import org.testng.annotations.Test;
+import org.testng.xml.XmlSuite;
+import org.testng.xml.XmlTest;
+import test.SimpleBaseTest;
+
+public class AfterGroupsTest extends SimpleBaseTest {
+
+  @Test(dataProvider = "dp")
+  public void ensureAfterGroupsInvoked(Class<?> clazz, Supplier<Boolean> actual) {
+    XmlSuite suite = createXmlSuite("sample_suite");
+    XmlTest test = createXmlTest(suite, "sample_test", clazz);
+    test.addIncludedGroup("A");
+    TestNG testng = create(suite);
+    testng.run();
+    assertTrue(actual.get(), "AfterGroups should have been invoked");
```

Comment:
In fact, I think the after groups methods should not have been invoked in the failing test because `alwaysRun` is not set.

According to the documentation:
> For after methods (afterSuite, afterClass, ...): If set to true, this configuration method will be run even if one or more methods invoked previously failed or was skipped.

You should add 2 more tests: success with alwaysRun + error with alwaysRun.
And the behavior may change if the exception comes from the test method instead of the configuration method.

Reply:
> In fact, I think the after groups methods should not have been invoked in the failing test because `alwaysRun` is not set.

That is not correct. The documentation is not applicable for `@AfterGroups` annotation.

> For before methods (beforeSuite, beforeTest, beforeTestClass and beforeTestMethod, but not beforeGroups): If set to true, this configuration method will be run regardless of what groups it belongs to. 
For after methods (afterSuite, afterClass, ...): If set to true, this configuration method will be run even if one or more methods invoked previously failed or was skipped.

Please see [here](http://static.javadoc.io/org.testng/testng/7.0.0-beta1/org/testng/annotations/AfterGroups.html#alwaysRun--)

`AfterGroups` has never worked for skipped methods for a very long time. Only now we are having it work. 

So I don't think `alwaysRun` has any role to play here.

Reply:
Yes, agree for `@BeforeGroups`, as explained in the documentation, because it doesn't have any sense.
But it is not so clear for `@AfterGroups` and I think it makes sense to have it here.

Reply:
I found an existing test that is already doing that. I have regrouped the tests into one single place so that its now evident. Please check now.

Reply:
@juherr - I think that the current set of tests is sufficient to validate this PR. I can add more tests separately but would prefer that this PR gets merged if there are no more outstanding comments to be addressed.

Reply:
In fact, I'm not sure about the fix and more tests will help me to see the full behavior.

Reply:
The changeset is to ensure that `@AfterGroups` without `alwasyRun=true` when 

1. There are test failures
2. There are test skips

This is what is being asked as part of #165 and I have added tests needed for the same along with the changeset.

Reply:
Ok, right.
But I have always some difficulties to ship something when I have doubts.